squirrel/static_analyser/analyser.cpp

```diff
@@ -1,62 +1,3581 @@
+  bool allReturns = true;
+
+  for (auto &c : swtch->cases()) {
+    allReturns &= checkNode(c.stmt);
+  }
+
+  if (swtch->defaultCase().stmt) {
+    allReturns &= checkNode(swtch->defaultCase().stmt);
+  }
+
+  return allReturns;
+}
+
+bool FunctionReturnTypeEvaluator::checkTry(const TryStatement *stmt) {
+  bool retTry = checkNode(stmt->tryStatement());
+  bool retCatch = checkNode(stmt->catchStatement());
+  return retTry && retCatch;
+}
+
+class PredicateCheckerVisitor : public Visitor {
+  bool deepCheck;
+  bool result;
+  const Node *checkee;
+protected:
+  NodeEqualChecker equalChecker;
+
+  PredicateCheckerVisitor(bool deep) : deepCheck(deep) {}
+
+  virtual bool doCheck(const Node *checkee, Node *n) const = 0;
+
+public:
+  void visitNode(Node *n) {
+    if (doCheck(checkee, n)) {
+      result = true;
+      return;
     }
+
+    if (deepCheck)
+      n->visitChildren(this);
   }
 
-  if (expr->op() == TO_OROR || expr->op() == TO_ANDAND) {
-    if (l->op() == TO_AND || l->op() == TO_OR || r->op() == TO_AND || r->op() == TO_OR) {
-      report(expr, DiagnosticsId::DI_BITWISE_BOOL_PAREN);
+  bool check(const Node *toCheck, Node *tree) {
+    result = false;
+    checkee = toCheck;
+    tree->visit(this);
+    return result;
+  }
+};
+
+class CheckModificationVisitor : public PredicateCheckerVisitor {
+protected:
+  bool doCheck(const Node *checkee, Node *n) const {
+    enum TreeOp op = n->op();
+
+    if (op == TO_ASSIGN || op == TO_INEXPR_ASSIGN || (TO_PLUSEQ <= op && op <= TO_MODEQ)) {
+      BinExpr *bin = static_cast<BinExpr *>(n);
+      return equalChecker.check(checkee, bin->lhs());
     }
+
+    if (op == TO_INC) {
+      IncExpr *inc = static_cast<IncExpr *>(n);
+      return equalChecker.check(checkee, inc->argument());
+    }
+
+    return false;
+  }
+public:
+  CheckModificationVisitor() : PredicateCheckerVisitor(false) {}
+};
+
+class ExistsChecker : public PredicateCheckerVisitor {
+protected:
+
+  bool doCheck(const Node *checkee, Node *n) const {
+    return equalChecker.check(checkee, n);
+  }
+
+public:
+
+  ExistsChecker() : PredicateCheckerVisitor(true) {}
+};
+
+class ModificationChecker : public Visitor {
+  bool result;
+public:
+
+  void visitNode(Node *n) {
+    enum TreeOp op = n->op();
+
+    switch (op)
+    {
+    case TO_INC:
+    case TO_ASSIGN:
+    case TO_INEXPR_ASSIGN:
+    case TO_PLUSEQ:
+    case TO_MINUSEQ:
+    case TO_MULEQ:
+    case TO_DIVEQ:
+    case TO_MODEQ:
+      result = true;
+      return;
+    default:
+      n->visitChildren(this);
+      break;
+    }
+  }
+
+  bool check(Node *n) {
+    result = false;
+    n->visit(this);
+    return result;
   }
+};
 
-  expr->visitChildren(this);
+static bool isBinaryArith(const Expr *expr) {
+  return TO_OROR <= expr->op() && expr->op() <= TO_SUB;
 }
 
+static bool isAssignExpr(const Expr *expr) {
+  enum TreeOp op = expr->op();
+  return op == TO_ASSIGN || op == TO_INEXPR_ASSIGN || (TO_PLUSEQ <= op && op <= TO_MODEQ);
```

Comment:
maybe you need to add comments to enum, that will say that you are relying on the fact that the new operators that will modify the variable in place should be in this range

Reply:
Or maybe named entries like `TO_OPERATOR_START` and `TO_OPERATOR_END`

Reply:
Add note to sqast.h